File: backtester/plan_date_limits.py
"""Tier-based backtest date windows (mirrors apps/web/src/lib/planDateLimits.ts)."""

from __future__ import annotations

from datetime import date, datetime, timedelta, timezone

TRIAL_HISTORY_DAYS = 90
PAID_HISTORY_DAYS = 1095

TIER_MAX_DAYS: dict[str, int] = {
    "trial": TRIAL_HISTORY_DAYS,
    "basic": PAID_HISTORY_DAYS,
    "starter": PAID_HISTORY_DAYS,
    "pro": PAID_HISTORY_DAYS,
    "automatic": PAID_HISTORY_DAYS,
}


def _parse_iso(value: str | None) -> date | None:
    if not value:
        return None
    return date.fromisoformat(value[:10])


def tier_history_window(
    tier: str,
    registered_at: str | None = None,
    today: date | None = None,
) -> tuple[str, str]:
    today = today or datetime.now(timezone.utc).date()
    max_days = TIER_MAX_DAYS.get(tier, TRIAL_HISTORY_DAYS)

    if tier == "trial":
        anchor = _parse_iso(registered_at) or today
        start = anchor - timedelta(days=max_days)
        return start.isoformat(), today.isoformat()

    start = today - timedelta(days=max_days)
    return start.isoformat(), today.isoformat()


def clamp_coin_range_to_plan(
    data_from: str,
    data_to: str,
    window_from: str,
    window_to: str,
) -> tuple[str, str] | None:
    start = max(data_from, window_from)
    end = min(data_to, window_to)
    if start > end:
        return None
    return start, end


def validate_backtest_date_range(
    start: str,
    end: str,
    coverage_from: str,
    coverage_to: str,
) -> str | None:
    if start > end:
        return "Начало периода позже конца"
    if start < coverage_from or end > coverage_to:
        return f"Период вне доступного диапазона ({coverage_from} — {coverage_to})"
    return None

File: apps/web/src/lib/planDateLimits.ts
import type { PlanTier } from "@/lib/subscriptionTiers";
import { featuresForTier } from "@/lib/subscriptionTiers";

/** Календарные дни истории на пробном тарифе (≈3 месяца). */
export const TRIAL_HISTORY_DAYS = 90;

/** Календарные дни истории на Basic / Pro (3 года). */
export const PAID_HISTORY_DAYS = 1095;

export interface PlanDateContext {
  tier: PlanTier;
  /** YYYY-MM-DD — дата регистрации (UTC), якорь пробного окна */
  registeredAt: string | null;
}

export function isoDateUtc(d: Date = new Date()): string {
  return d.toISOString().slice(0, 10);
}

export function addCalendarDays(iso: string, delta: number): string {
  const d = new Date(iso + "T12:00:00Z");
  d.setUTCDate(d.getUTCDate() + delta);
  return isoDateUtc(d);
}

export function maxIso(a: string, b: string): string {
  return a > b ? a : b;
}

export function minIso(a: string, b: string): string {
  return a < b ? a : b;
}

/** Доступное окно дат по тарифу (до пересечения с данными монеты). */
export function tierHistoryWindow(
  ctx: PlanDateContext,
  today: string = isoDateUtc()
): { from: string; to: string } {
  const maxDays = featuresForTier(ctx.tier).maxDays;

  if (ctx.tier === "trial") {
    const anchor = ctx.registeredAt ?? today;
    return {
      from: addCalendarDays(anchor, -maxDays),
      to: today,
    };
  }

  return {
    from: addCalendarDays(today, -maxDays),
    to: today,
  };
}

/** Пересечение свечей монеты с тарифным окном. */
export function clampCoinRangeToPlan(
  dataFrom: string,
  dataTo: string,
  window: { from: string; to: string }
): { from: string; to: string } | null {
  const from = maxIso(dataFrom, window.from);
  const to = minIso(dataTo, window.to);
  if (from > to) return null;
  return { from, to };
}

export function validateBacktestDateRange(
  start: string,
  end: string,
  coverage: { from: string; to: string }
): string | null {
  if (start > end) return "Начало периода позже конца";
  if (start < coverage.from || end > coverage.to) {
    return `Период вне доступного диапазона (${coverage.from} — ${coverage.to})`;
  }
  return null;
}
